Load parent package config

sigurdm opened the pull request:
Fixes #150196

sigurdm committed:
Refactor BuildInfo to always require packageConfigPath

sigurdm committed:
Address review

sigurdm committed:
Fix argument

sigurdm committed:
Remove unused import

sigurdm committed:
Merge branch 'master' into make_build_info_package_config_required

sigurdm committed:
Search for package config in parent directories

sigurdm committed:
Merge remote-tracking branch 'origin/master' into make_build_info_package_config_required

sigurdm committed:
formatting

sigurdm committed:
formatting

sigurdm committed:
test pub get

sigurdm committed:
Dont import package:path

andrewkolos commented:
Looks like this test is failing: https://github.com/flutter/flutter/blob/0a226c328e2941f86002250edc55e03400158d0d/packages/flutter_tools/test/commands.shard/permeable/packages_test.dart#L320

andrewkolos commented on packages/flutter_tools/lib/src/dart/package_map.dart:
```diff
@@ -16,6 +16,44 @@ Future<PackageConfig> currentPackageConfig() async {
+    }
+    final String nextDir = fileSystem.path.dirname(candidateDir);
+    if (nextDir == candidateDir) {
+      return null;
+    }
+    candidateDir = nextDir;
+  }
```
Nit: I think this would be slightly more readable.

```suggestion
    final String parentDir = fileSystem.path.dirname(candidateDir);
    
    if (fileSystem.identicalSync(parentDir, candidateDir)) {
      return null;
    }
    candidateDir = parentDir;
```

andrewkolos commented on packages/flutter_tools/lib/src/dart/package_map.dart:
```diff
@@ -16,6 +16,44 @@ Future<PackageConfig> currentPackageConfig() async {
+    if (fileSystem.file(candidatePackageConfigFile).existsSync()) {
+      return candidatePackageConfigFile;
+    }
+    // TODO(sigurdm): we should not need to check this file, it is obsolete.
+    final File candidatePackagesFile = fileSystem.file(fileSystem.path.join(candidateDir, '.packages'));
+    if (fileSystem.file(candidatePackagesFile).existsSync()) {
+      return candidatePackagesFile;
+    }
```
If .packages is obsolete? Can we just remove this? If not, consider filing a follow-up issue and linking it here (with P3 and tech-debt labels).

andrewkolos commented on packages/flutter_tools/lib/src/dart/package_map.dart:
```diff
@@ -16,6 +16,44 @@ Future<PackageConfig> currentPackageConfig() async {
+    if (fileSystem.file(candidatePackagesFile).existsSync()) {
+      return candidatePackagesFile;
+    }
+    final String nextDir = fileSystem.path.dirname(candidateDir);
+    if (nextDir == candidateDir) {
+      return null;
+    }
+    candidateDir = nextDir;
```
Nit: I think this would be slightly more readable.
```suggestion
    final String parentDir = fileSystem.path.dirname(candidateDir);
    if (fileSystem.identicalSync(parentDir, candidateDir)) {
      return null;
    }
    candidateDir = parentDir;
```

sigurdm committed:
Fix test

sigurdm committed:
Reinstate .packages check

sigurdm committed:
Use issue link in TODO

sigurdm committed:
TODO style

andrewkolos committed:
Merge branch 'master' into pr/sigurdm/150850-1

engine-flutter-autoroll committed:
Roll Flutter Engine from 0d93da7e2fe1 to b57a044ed10f (1 revision) (#151082)

https://github.com/flutter/engine/compare/0d93da7e2fe1...b57a044ed10f

2024-07-01 skia-flutter-autoroll@skia.org Roll Dart SDK from ab6a48438e07 to 338c6d4fd9c5 (1 revision) (flutter/engine#53658)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jimgraham@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

jason-simmons committed:
Fix teardown of a FocusScopeNode in material/app_test (#151115)

This test suite had been failing when run with --test-randomize-ordering-seed=20240701

flutter-pub-roller-bot committed:
Roll pub packages (#151028)

This PR was generated by `flutter update-packages --force-upgrade`.

Pante committed:
Fix missing `[` in docs (#151091)

This is a trivial fix for a typo in the docs.

goderbauer committed:
Bump dartdoc to 8.0.10 (#151107)

For https://github.com/flutter/flutter/pull/151011#discussion_r1660046408.

goderbauer committed:
More docimports for animation library (#151011)

docimports for references in doc comments on library statements now work!

Part of https://github.com/flutter/flutter/issues/150800

jtmcdole committed:
Read AndroidManifest.xml and emit manifest-aar-impeller-(enabled|disabled) analytics (#150970)

Same as #150791 except with AARs (`flutter build module`)

engine-flutter-autoroll committed:
Roll Flutter Engine from b57a044ed10f to 3456fee1a6b9 (5 revisions) (#151127)

https://github.com/flutter/engine/compare/b57a044ed10f...3456fee1a6b9

2024-07-01 skia-flutter-autoroll@skia.org Roll Skia from a67a67b7452b to a62bf018429c (2 revisions) (flutter/engine#53665)
2024-07-01 skia-flutter-autoroll@skia.org Roll Skia from 8af866a56bce to a67a67b7452b (1 revision) (flutter/engine#53664)
2024-07-01 skia-flutter-autoroll@skia.org Roll Fuchsia Linux SDK from DSPuQ9O6WejkIBUUh... to LkXpxHsQlkPT4mmJ7... (flutter/engine#53663)
2024-07-01 skia-flutter-autoroll@skia.org Roll Skia from e27104eaef37 to 8af866a56bce (4 revisions) (flutter/engine#53661)
2024-07-01 skia-flutter-autoroll@skia.org Roll Skia from 34be7c830184 to e27104eaef37 (1 revision) (flutter/engine#53659)

Also rolling transitive DEPS:
  fuchsia/sdk/core/linux-amd64 from DSPuQ9O6Wejk to LkXpxHsQlkPT

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jimgraham@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

gmackall committed:
Upgrade template Gradle, App AGP, Module AGP, and Kotlin versions, and tests (#150969)

After the land of https://github.com/flutter/engine/pull/53592, there is some log spam:
```
e: /Users/mackall/.gradle/caches/transforms-3/c1e137371ec1afe9bc9bd7b05823752d/transformed/fragment-1.7.1/jars/classes.jar!/META-INF/fragment_release.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: /Users/mackall/.gradle/caches/transforms-3/d86c7cb1c556fe1655fa56db671c649c/transformed/jetified-activity-1.8.1/jars/classes.jar!/META-INF/activity_release.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
...
```

I think this is harmless, but still annoying. Upgrading the AGP version fixes it. To be honest, I don't know why (I expected the Kotlin version would do it). But after https://github.com/flutter/flutter/pull/146307, our tests have been running on AGP/Gradle 8.1/8.3 for a while, so it makes sense to upgrade anyways.

In a follow up PR:
1. Also upgrade the tests that were left behind in https://github.com/flutter/flutter/pull/146307, as I think removal of discontinued plugins paved the way here.

auto-submit[bot] committed:
Reverts "Upgrade template Gradle, App AGP, Module AGP, and Kotlin versions, and tests (#150969)" (#151147)

Reverts: flutter/flutter#150969
Initiated by: goderbauer
Reason for reverting: Failing test in https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8743574743030691569/+/u/run_android_obfuscate_test/stdout
Original PR Author: gmackall

Reviewed By: {christopherfujino, reidbaker}

This change reverts the following previous change:
After the land of https://github.com/flutter/engine/pull/53592, there is some log spam:
```
e: /Users/mackall/.gradle/caches/transforms-3/c1e137371ec1afe9bc9bd7b05823752d/transformed/fragment-1.7.1/jars/classes.jar!/META-INF/fragment_release.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: /Users/mackall/.gradle/caches/transforms-3/d86c7cb1c556fe1655fa56db671c649c/transformed/jetified-activity-1.8.1/jars/classes.jar!/META-INF/activity_release.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
...
```

I think this is harmless, but still annoying. Upgrading the AGP version fixes it. To be honest, I don't know why (I expected the Kotlin version would do it). But after https://github.com/flutter/flutter/pull/146307, our tests have been running on AGP/Gradle 8.1/8.3 for a while, so it makes sense to upgrade anyways.

In a follow up PR:
1. Also upgrade the tests that were left behind in https://github.com/flutter/flutter/pull/146307, as I think removal of discontinued plugins paved the way here.

nate-thegrate committed:
Implementing a few switch statements (#150946)

I really like how patterns can be used for variable assignment and avoiding duplicated logic. (related: #150942)

```dart
// before
final GestureRecognizer? recognizer = info.recognizer;
if (recognizer is TapGestureRecognizer) {
  if (recognizer.onTap != null) {
    configuration.onTap = recognizer.onTap;
    configuration.isLink = true;
  }
} else if (recognizer is DoubleTapGestureRecognizer) {
  if (recognizer.onDoubleTap != null) {
    configuration.onTap = recognizer.onDoubleTap;
    configuration.isLink = true;
  }
}

// after
switch (info.recognizer) {
  case TapGestureRecognizer(:final VoidCallback? onTap):
  case DoubleTapGestureRecognizer(onDoubleTap: final VoidCallback? onTap):
    if (onTap != null) {
      configuration.onTap = onTap;
      configuration.isLink = true;
    }
}
```

andrewkolos committed:
[tool] remove some temporary printTrace calls (#151074)

Cleans up printTrace calls added while investigating https://github.com/flutter/flutter/issues/146879.

Also makes[ this `catch` clause ](https://github.com/flutter/flutter/pull/151074/files#diff-214723283174eb75e740a102252215723aa56716226ea92d084ef8a5bea4e931L578-R581)more precise.

engine-flutter-autoroll committed:
Roll Flutter Engine from 3456fee1a6b9 to fc5bc14e6091 (8 revisions) (#151150)

https://github.com/flutter/engine/compare/3456fee1a6b9...fc5bc14e6091

2024-07-01 98614782+auto-submit[bot]@users.noreply.github.com Reverts "Output .js files as ES6 modules. (#52023)" (flutter/engine#53674)
2024-07-01 skia-flutter-autoroll@skia.org Roll Dart SDK from 21c08743ee4a to c23e58143793 (1 revision) (flutter/engine#53670)
2024-07-01 skia-flutter-autoroll@skia.org Roll Skia from ae2b97d74812 to 8375bdc6e191 (3 revisions) (flutter/engine#53669)
2024-07-01 jacksongardner@google.com Output .js files as ES6 modules. (flutter/engine#52023)
2024-07-01 skia-flutter-autoroll@skia.org Roll Skia from a62bf018429c to ae2b97d74812 (3 revisions) (flutter/engine#53668)
2024-07-01 fmil@google.com [icu] Ignores the dir `flutter/third_party/icu/patches` (flutter/engine#53667)
2024-07-01 jonahwilliams@google.com [Impeller] track the sizes of all outstanding MTLTexture allocations and report per frame in MB, matching Vulkan implementation. (flutter/engine#53618)
2024-07-01 skia-flutter-autoroll@skia.org Roll Dart SDK from 338c6d4fd9c5 to 21c08743ee4a (1 revision) (flutter/engine#53666)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jimgraham@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

dkwingsmt committed:
Fix: `CupertinoActionSheet` should take up max height when actions section is short (#150708)

This PR addresses a bug from the previous rewrite of the `CupertinoActionSheet` height allocation algorithm. 

The previous approach assigns the content section with a minimal height. As a result, if the actions section is shorter than `_kActionSheetActionsSectionMinHeight`, the remaining space will not be taken by the content section.

As far as I know, this algorithm can not be implemented by simply compositing simple layout widgets, and therefore I created a new widget `_PriorityColumn` that uses a custom layout algorithm. (The resulting code is still shorter (and cleaner) than before the rewrite!)

Some other trivial refactor is also done to clean up the code around parameters of `_ActionSheetMainSheet`.

engine-flutter-autoroll committed:
Roll Flutter Engine from fc5bc14e6091 to d3c5bd66a78f (1 revision) (#151155)

https://github.com/flutter/engine/compare/fc5bc14e6091...d3c5bd66a78f

2024-07-02 skia-flutter-autoroll@skia.org Manual roll ICU from 98f2494518c2 to 43953f57b037 (1 revision) (flutter/engine#53675)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jimgraham@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

engine-flutter-autoroll committed:
Roll Flutter Engine from d3c5bd66a78f to 40c087b31515 (1 revision) (#151156)

https://github.com/flutter/engine/compare/d3c5bd66a78f...40c087b31515

2024-07-02 robert.ancell@canonical.com Restore creation of engine before Linux widget is realized. (flutter/engine#53604)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jimgraham@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

sigurdm committed:
Refactor BuildInfo to always require packageConfigPath (#150559)

Refactor warming up to #150196

sigurdm committed:
Fix merge

sigurdm committed:
Merge

sigurdm committed:
Remove debug print

sigurdm committed:
Use correct packages path for _updateDevFS

sigurdm committed:
Handle building and assembling

andrewkolos commented:
Is this ready for review? I see failing checks, but I don't know if these are flakes

sigurdm committed:
merge

sigurdm committed:
Only store packageConfigPath in Environment

sigurdm committed:
Merge remote-tracking branch 'origin/master' into load_parent_package_config

sigurdm commented:
I think all flutter tests are passing now. And the PR can be reviewed - but we still need a G3 hotfix.

Will work on that tomorrow.

sigurdm committed:
Merge remote-tracking branch 'origin/master' into load_parent_package_config

sigurdm commented:
And now it seems we're all green!

sigurdm commented:
PTAL

mit-mit commented:
@sigurdm looks like you need to update/rebase this?

sigurdm committed:
Merge remote-tracking branch 'origin/master' into load_parent_package_config

sigurdm committed:
Merge branch 'master' into load_parent_package_config

andrewkolos commented on packages/flutter_tools/lib/src/build_system/build_system.dart:
```diff
@@ -371,9 +372,11 @@ class Environment {
 
     final Directory rootBuildDir = buildDir ?? projectDir.childDirectory('build');
     final Directory buildDirectory = rootBuildDir.childDirectory(buildPrefix);
+    // final File packageConfigFile = findPackageConfigFileOrDefault(projectDir);
```
```suggestion
```

andrewkolos commented on packages/flutter_tools/lib/src/build_system/source.dart:
```diff
@@ -99,6 +99,12 @@ class SourceVisitor implements ResolvedFiles {
+          environment.fileSystem.path.dirname(
+            environment.fileSystem.path.dirname(
+              environment.packageConfigPath,
+            )
```
turbo nit:
```suggestion
            ),
```

sigurdm committed:
Update packages/flutter_tools/lib/src/build_system/build_system.dart

Co-authored-by: Andrew Kolos <andrewrkolos@gmail.com>

sigurdm committed:
Update packages/flutter_tools/lib/src/build_system/source.dart

Co-authored-by: Andrew Kolos <andrewrkolos@gmail.com>

sigurdm committed:
Merge branch 'master' into load_parent_package_config

sigurdm committed:
Merge branch 'master' into load_parent_package_config

sigurdm committed:
Merge branch 'master' into load_parent_package_config

christopherfujino commented on packages/flutter_tools/lib/src/build_system/build_system.dart:
```diff
@@ -457,6 +462,10 @@ class Environment {
   static const String kProjectDirectory = '{PROJECT_DIR}';
 
+  /// The [Source] value which is substituted with the path to the parent of
+  /// [buildInfo.packageConfigPath].
```
nit, if you make this `[BuildInfo.packageConfigPath]`, editor jump to definition should work.

sigurdm replied:
Well - Should really be "grandparent" i guess. 
Tried to clarify

christopherfujino commented on packages/flutter_tools/lib/src/build_system/build_system.dart:
```diff
@@ -457,6 +462,10 @@ class Environment {
 
+  /// The [Source] value which is substituted with the path to the parent of
+  /// [buildInfo.packageConfigPath].
+  static const String kWorkspaceDirectory = '{WORKSPACE_DIR}';
```
so, in practice, `{WORKSPACE_DIR}` == `.dart_tool/`?

sigurdm replied:
Well - Should really be "grandparent" i guess. 
Tried to clarify

christopherfujino commented on packages/flutter_tools/lib/src/build_system/source.dart:
```diff
@@ -99,6 +99,12 @@ class SourceVisitor implements ResolvedFiles {
         // flutter root will not contain a symbolic link.
         Environment.kFlutterRootDirectory => environment.flutterRootDir.absolute.path,
         Environment.kProjectDirectory     => environment.projectDir.resolveSymbolicLinksSync(),
+        Environment.kWorkspaceDirectory   =>
```
oh, so the workspace dir is actually the PARENT of `.dart_tool/`?

sigurdm replied:
Yeah - I'll try to clarify!

christopherfujino commented on packages/flutter_tools/lib/src/dart/package_map.dart:
```diff
@@ -16,6 +16,45 @@ Future<PackageConfig> currentPackageConfig() async {
+///
+/// Returns `null` if no package_config.json was found.
+File? findPackageConfigFile(Directory dir) {
+  final FileSystem fileSystem = dir.fileSystem;
```
nit, this can be a little cleaner by using more of the built-in APIs from package:file:

```dart
File? findPackageConfigFile(Directory dir) {
  final FileSystem fileSystem = dir.fileSystem;

  Directory candidateDir = fileSystem.directory(dir.path).absolute;
  while (true) {
    final File candidatePackageConfigFile = candidateDir
        .childDirectory('.dart_tool')
        .childFile('package_config.json');
    if (candidatePackageConfigFile.existsSync()) {
      return candidatePackageConfigFile;
    }
    // TODO(sigurdm): we should not need to check this file, it is obsolete,
    // https://github.com/flutter/flutter/issues/150908.
    final File candidatePackagesFile = candidateDir.childFile('.packages');
    if (candidatePackagesFile.existsSync()) {
      return candidatePackagesFile;
    }
    final Directory parentDir = candidateDir.parent;
    if (fileSystem.identicalSync(parentDir.path, candidateDir.path)) {
      return null;
    }
    candidateDir = parentDir;
  }
}
```

sigurdm replied:
Nice!

christopherfujino commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -694,7 +694,7 @@ class TestCommand extends FlutterCommand with DeviceBasedDevelopmentArtifacts {
     final AssetBundle assetBundle = AssetBundleFactory.instance.createBundle();
     final int build = await assetBundle.build(
-      packagesPath: '.packages',
+      packageConfigPath: '.dart_tool/package_config.json',
```
why do we hard-code here? do we not support workspaces in tests?

sigurdm replied:
I guess just to make existing tests continue working.
If we want to make many workspace tests it could totally be parametrized.

sigurdm replied:
Done

christopherfujino commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -258,50 +258,87 @@ class _DefaultPub implements Pub {
+    // finding .dart_tool/package_config.json we will not mistakenly find a
+    // package_config.json from outside the workspace.
+    //
+    // TODO(sigurdm): avoid relying on pubs private parts somehow?
```
nit I'd use the phrase "private implementation details" over "private parts"

christopherfujino commented on packages/flutter_tools/lib/src/build_system/targets/native_assets.dart:
```diff
@@ -64,17 +64,15 @@ class NativeAssets extends Target {
+
       final PackageConfig packageConfig = await loadPackageConfigWithLogging(
-        packagesFile,
+        environment.fileSystem.file(environment.packageConfigPath),
```
nit looks like `fileSystem` is already a local variable here that's equivalent to `environment.fileSystem`

sigurdm replied:
Thanks!

christopherfujino commented on packages/flutter_tools/lib/src/dart/package_map.dart:
```diff
@@ -16,6 +16,45 @@ Future<PackageConfig> currentPackageConfig() async {
+///
+/// Like [findPackageConfigFile] but returns
+/// `$dir/.dart_tool/package_config.json` if no package config could be found.
+File findPackageConfigFileOrDefault(Directory dir) {
```
It seems like we call this function a lot, is there any reason we can't cache it as a field in the `Environment` class?

sigurdm replied:
Yeah - this is a good point.

If we ensure that it is only called after pub get, then it should be fine.

I guess we are caching it in `BuildInfo` - everywhere should just read it from there.

Added a TODO

christopherfujino commented on packages/flutter_tools/lib/src/commands/assemble.dart:
```diff
@@ -217,10 +217,10 @@ class AssembleCommand extends FlutterCommand {
 
   /// The environmental configuration for a build invocation.
-  Environment _createEnvironment() {
+  Future<Environment> _createEnvironment() async {
```
Why did this become async?

sigurdm replied:
I think at some point while I developed this it needed something async - seems that is no longer the case.

Made sync again.

christopherfujino commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -258,50 +258,87 @@ class _DefaultPub implements Pub {
+          pubspecYaml.lastModifiedSync().isBefore(packageConfigFile.lastModifiedSync()) &&
+          lastVersion.existsSync() &&
+          lastVersion.readAsStringSync() == currentVersion.readAsStringSync()) {
       _logger.printTrace('Skipping pub get: version match.');
```
I think this should be indented 2 spaces to the right.

christopherfujino commented on packages/flutter_tools/lib/src/dart/pub.dart:
```diff
@@ -688,6 +729,10 @@ class _DefaultPub implements Pub {
     }
+
+    if (!_fileSystem.path.equals(packageConfigFile.parent.parent.path, project.directory.path)) {
+      throwToolExit('`generate: true` is not supported in with workspaces.');
```
nit: s/in with/within

sigurdm replied:
Done

christopherfujino commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -1750,21 +1768,22 @@ Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and
         checkUpToDate: cachePubGet,
       );
 
+      await generateLocalizationsSyntheticPackage(
```
why did you move this after the pub.get? I would think that we would want to pub get after the synthetic package is created?

sigurdm replied:
It depends on `environment` that now contains the package config path. That is not valid before pub get has run.
Also I don't see what pub has to use the synthetic package for...

christopherfujino replied:
i trust you :)

sigurdm replied:
I really don't have a good feel for the generateLocalizationsSyntheticPackage mechanism. But I don't believe it interacts with package resolution besides creating a package that is injected into package_config.json. So fingers crossed the order should not really matter.

christopherfujino replied:
My naive understanding was that the project's package_config.json would need to have a reference to the synthetic package, and thus it would need to exist so that pub would find it and write it to package_config.json.

@jonahwilliams we might need your help here...

sigurdm replied:
pub itself will never write the synthetic package into package config. That is done by flutter_tool here:
https://github.com/flutter/flutter/blob/aaab2e145f20c87c27b39b7c4a382733a420e8fc/packages/flutter_tools/lib/src/dart/pub.dart#L751

That happens at the end of `flutter pub get`: https://github.com/flutter/flutter/blob/aaab2e145f20c87c27b39b7c4a382733a420e8fc/packages/flutter_tools/lib/src/dart/pub.dart#L369

christopherfujino replied:
ohhhh, yeah, I forgot about that. :)

christopherfujino commented:
Mostly nits, a few questions, but I think we can get this merged soon.

sigurdm committed:
Address easy comments

christopherfujino reviewed (approved):
This LGTM if you add the TODO to [packages/flutter_tools/lib/src/commands/test.dart](https://github.com/flutter/flutter/pull/150850/files/2fa2ae2c525be708c52f3cfaad28a915553b6a01#diff-328d8a42e42b84152d577569aa043fdfa16977190165f1a11897f6f47d81c1a5)

andrewkolos reviewed (approved):
LGTM. Sorry for the silence.

sigurdm committed:
Add TODO
